atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/store/UnreliableTimestampManagerTest.java

```diff
@@ -0,0 +1,170 @@
+        }
+
+        assertListRandomlyIncreasesInValue(allValues);
+    }
+
+    @Test
+    public void randomlyIncreaseTimestampRandomlyIncreasesTimestampWithMultipleThreads() throws InterruptedException {
+        ArrayList<Long> allValues = new ArrayList<>(List.of(0L));
+        UnreliableTimestampManager timestampManager = createThreadsafeArrayBackedTimestampManager(allValues);
+
+        ExecutorService executor = Executors.newCachedThreadPool();
+        try {
+            for (int i = 0; i < 10; i++) {
+                executor.execute(timestampManager::randomlyIncreaseTimestamp);
+            }
+        } finally {
+            executor.shutdown();
+            assertThat(executor.awaitTermination(10, TimeUnit.SECONDS)).isTrue();
+        }
+
+        assertListRandomlyIncreasesInValue(allValues);
+    }
+
+    @ParameterizedTest
+    @MethodSource("timestampMethods")
+    public void gettingTimestampsBlocksOnPendingFastForward(Consumer<TimestampManager> task)
+            throws InterruptedException {
+        CountDownLatch startingFastForward = new CountDownLatch(1);
+        CountDownLatch blockFastForward = new CountDownLatch(1);
+
+        UnreliableTimestampManager timestampManager =
+                UnreliableTimestampManager.create(timelockService, () -> 0L, _value -> {
+                    startingFastForward.countDown();
+                    try {
+                        blockFastForward.await();
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+
+        ExecutorService executor = Executors.newCachedThreadPool();
+        try {
+            Future<?> increaseTimestampFuture = executor.submit(timestampManager::randomlyIncreaseTimestamp);
+            startingFastForward.await();
+            Future<?> getTimestamp = executor.submit(() -> task.accept(timestampManager));
+            assertThat(getTimestamp).isNotDone();
+            blockFastForward.countDown();
+            assertThat(increaseTimestampFuture).succeedsWithin(Duration.ofSeconds(1));
+            assertThat(getTimestamp).succeedsWithin(Duration.ofSeconds(1));
+
+        } finally {
+            executor.shutdown();
+            boolean shutdownSuccessfully = executor.awaitTermination(10, TimeUnit.SECONDS);
+            executor.shutdownNow();
+            assertThat(shutdownSuccessfully).isTrue();
+        }
+    }
+
+    private static void assertListRandomlyIncreasesInValue(List<Long> timestamps) {
+        assertThat(timestamps).isSorted();
+        assertThat(new HashSet<>(timestamps)).hasSameSizeAs(timestamps);
+
+        List<Long> differencesBetweenValues = new ArrayList<>();
+        for (int i = 1; i < timestamps.size(); i++) {
+            differencesBetweenValues.add(timestamps.get(i) - timestamps.get(i - 1));
+        }
+
+        // Verifying things increase randomly. We can't assert that every delta is random, since it's possible that at
```

Comment:
consider: I am not a huge fan of verifying randomness like this. I agree flaking is unlikely here with things as-is (though this might change if others use your method in other ways with data generated randomly but gets transformed/narrowed down to have lower entroy). Ideally, we'd make SecureRandom (or whatever supplier we need) a dependency but I do not feel strongly about this.

Reply:
Discussed, we'll leave abstracting out the randomness across the antithesis tests to another PR